src/fsharp/CompileOps.fs

```diff
@@ -2549,6 +2547,56 @@ let GetAutoOpenAttributes ilg ilModule =
 
 let GetInternalsVisibleToAttributes ilg ilModule = 
     ilModule |> GetCustomAttributesOfILModule |> List.choose (TryFindInternalsVisibleToAttr ilg)
+
+let findPrimaryAssembly (data: TcConfigBuilder) =
+    let assemblies =
+        data.referencedDLLs
+        |> List.map (fun r -> r.Text)
+
+    let readerSettings: ILReaderOptions = 
+        { pdbDirPath=None
+          ilGlobals = EcmaMscorlibILGlobals
+          reduceMemoryUsage = data.reduceMemoryUsage
+          metadataOnly = MetadataOnlyFlag.Yes
+          tryGetMetadataSnapshot = data.tryGetMetadataSnapshot }
+
+    let nameSystem = ["System"]
+    let nameObject = "Object"
+
+    let hasNoPiaLocalTypes (reader: ILModuleReader) =
+        reader.ILModuleDef.TypeDefs.AsArray
+        |> Array.exists (fun (ilTypeDef: ILTypeDef) ->
+            ilTypeDef.CustomAttrs.AsArray
+            |> Array.exists (fun attr -> attr.Method.MethodRef.DeclaringTypeRef.Name = "System.Runtime.InteropServices.TypeIdentifierAttribute"))
+
+    let hasBaselessObjectClass (reader: ILModuleReader) =
+        reader.ILModuleDef.TypeDefs.AsArrayOfPreTypeDefs
+        |> Array.exists (fun (ilPreTypeDef: ILPreTypeDef) ->
+            if ilPreTypeDef.Name = nameObject && ilPreTypeDef.Namespace = nameSystem then
+                let ilTypeDef = ilPreTypeDef.GetTypeDef()
+                ilTypeDef.IsClass && ilTypeDef.Implements.IsEmpty
+            else
+                false)
+
+    let isCandidate (reader: ILModuleReader) =
+        reader.ILAssemblyRefs.IsEmpty &&
+        hasBaselessObjectClass reader &&
+        not (hasNoPiaLocalTypes reader)
+
+    let candidates =
+        assemblies
+        |> List.choose (fun path ->
+            let reader = OpenILModuleReader path readerSettings
+            if isCandidate reader && reader.ILModuleDef.HasManifest then
```

Comment:
It may be faster to change order since isCandidate may be expensive

Reply:
Maybe. Most assemblies have a manifest so `isCandidate` will always get called, but it doesn't hurt to change the order.

Reply:
My hypothesis on `isCandidate`, it should be very quick on all assemblies because it first checks to see if the assembly uses any other assemblies. Which almost all assemblies do, except for the primary one.

Reply:
so in a refs directory (ie dotnet-root/packs/Microsoft.NETCore.App.Ref/3.1.0/ref/netcoreapp3.1), there are two dlls that I'd consider a 'primary assembly': mscorlib.dll and netstandard.dll. do both of those pass the candidate checks here? if so, then we'll always hit the multiple-candidates stage on line 2598.

Reply:
It's fine we that we have multiple candidates; it just shortens the list. The one we actually pick to be primary is the one that contains the type definition of `System.Object`.

Reply:
Hmm, I read my code incorrectly. In the case you have @baronfel , `isCandidate` should not return true for mscorlib.dll and netstandard.dll inside of that directory because they do not contain the actual type definition for `System.Object`.